### PR TITLE
change to more user-friendly return message if the cert already exists in proxy

### PR DESCRIPTION
**Pull request by tingyuz**
This is to fix isssue #2089 . Update user-friendly log message if the certificate already exists in reverse proxy.

Signed-off-by: Tingyu Zeng <Tingyu.Zeng@rsa.com>

### CODE REVIEW

**Comment by tingyuz** [comment]
recheck

File: internal/security/proxy/service.go [hunk 1]
```diff
@@ -173,8 +173,12 @@ func (s *Service) postCert(cert CertificateLoader) error {
 		if err != nil {
 			return err
 		}
-		e := fmt.Sprintf("failed to add certificate with errorcode %d, error %s", resp.StatusCode, string(b))
+		reason := string(b)
+		e := fmt.Sprintf("failed to add certificate with errorcode %d, error %s", resp.StatusCode, reason)
 		LoggingClient.Error(e)
+		if (resp.StatusCode == http.StatusBadRequest) && (strings.Index(reason, "existing certificate") != -1) {
+			e = fmt.Sprintf("certificate already exists on reverse proxy")
```
Comment (jim-wang-intel) [inline internal/security/proxy/service.go]: @tingyuz why the http status code not be `http.StatusConflict` when the certificate already exists?

Comment (anonymouse64) [reply internal/security/proxy/service.go]: FWIW the docs don't seem to say what happens when there's already a cert uploaded: https://docs.konghq.com/1.0.x/admin-api/#create-certificate

Comment (tingyuz) [reply internal/security/proxy/service.go]: Please refer my comments on #2089. The original code even checks http.StatusConflict but kong doesn't follow such standard, so we have to get this done this way.

**Comment by bnevis-i** [comment]
recheck

**Comment by bnevis-i** [comment]
The default clause of the following switch statement continues to have no code coverage:
```go
        switch resp.StatusCode {
        case http.StatusOK, http.StatusCreated, http.StatusConflict:
                LoggingClient.Info("successfully added certificate to the reverse proxy")
                break
        default:
                b, err := ioutil.ReadAll(resp.Body)
                if err != nil {
                        return err, false
                }
                reason := string(b)
                e := fmt.Sprintf("failed to add certificate with errorcode %d, error %s", resp.StatusCode, reason)
                LoggingClient.Error(e)
                if (resp.StatusCode == http.StatusBadRequest) && (strings.Index(reason, "existing certificate") != -1) {
                        e = fmt.Sprintf("certificate already exists on reverse proxy")
                        return nil, true
                }
                return errors.New(e), false
        }
```

File: internal/security/proxy/service.go (hunk 1 above)
Comment (jim-wang-intel) [reply internal/security/proxy/service.go]: ok

**Comment by tingyuz** [comment]
Add the unit test case for the changes, which is, when the cert already exists on the proxy.

File: internal/security/proxy/service.go [hunk 2]
```diff
@@ -171,13 +175,18 @@ func (s *Service) postCert(cert CertificateLoader) error {
...
 		LoggingClient.Error(e)
-		return errors.New(e)
+		if (resp.StatusCode == http.StatusBadRequest) && (strings.Index(reason, "existing certificate") != -1) {
+			e = fmt.Sprintf("certificate already exists on reverse proxy")
+			return nil, true
+		}
+		return errors.New(e), false
 	}
-	return nil
+	return nil, false
```
Comment (jim-wang-intel) [inline internal/security/proxy/service.go]: should return `true` here, too.
```suggestion
	return nil, true
```

Comment (bnevis-i) [reply internal/security/proxy/service.go]: This boolean means "existed" not "was successful"

Comment (jim-wang-intel) [reply internal/security/proxy/service.go]: ok, then please ignore my comment then.

**Comment by bnevis-i** [comment]
Note that we need to pull this fix into master too.

**Comment by mkbhanda** [comment]
To Jim's comment and your response Bryon, it indicates need for better self-documenting code. 
Go supports named return values, see https://tour.golang.org/basics/7
Worth editing.

**Comment by bnevis-i** [comment]
While we're on the the subject of the bad signature for postCert(), generally the error  is the last item in the return list.  I generally dislike returning information in a bool.  Moreover, for this function, there are multiple possible outcomes:  there was an error, it succeeded replacing the cert, or succeeded uploading the cert for the first time.  It may not be unreasonable that a future outcome might be, certs exists but was not replaced.  As such, I think that postCert() should just return an error, and it should take a pointer argument that returns a information enum that tells how it was handled if the function succeeded.

**Comment by anonymouse64** [comment]
+1

**Comment by tingyuz** [comment]
recheck

**Pull request merged**